### PR TITLE
New `spack.environment.active_environment` api, and make spack.environment not depend on spack.cmd.

### PR DESCRIPTION
**Pull request by haampie**
- Rework `spack.environment.get_env(...)` to `spack.environment.active_environment()`
  which is now a trivial getter for the active environment.
- Introduce `spack.cmd.require_active_env(cmd_name=...)` which is a wrapper 
  around `spack.environment.active_environment()` but errors with a helpful
  error message when used from the CLI.
- Clean up instances where former `spack.environment.get_env` was abused
outside the context of a command (fake command name, fake `-e` flag, 
  bunch of `None`, `None` arguments)
- Remove the `-e` parsing from former `get_env`, because `main.py` is
responsible for processing `-e` and already activates the environment.
  In fact this logic was entirely broken, because it did not respect `-D` at all
  for loading an env from a directory, and it did not respect `-E` (i.e. it would
  load an environment even when explicitly disabled).
- Move `spack.environment.find_environment` to
`spack.cmd.find_environment`, to avoid that spack.environment is aware
of argparse.
- Move command line argument parsing of environment `installs_specs` from 
  `spack.environment` to `spack.cmd`, to avoid that spack.environment is aware
  of argparse.
- Remove some unused imports, and work around circular imports.
- Use `ev` alias for `spack.environment` consistently where it works Python 2.7
  permitting

TL;DR: better interface for getting the currently active environment, and refactor
things so that `spack.environment` does not depend on `spack.cmd`.

### CODE REVIEW

**Comment by haampie** [comment]
Added the `breaking-change` label for posterity because `spack.environment.get_env` got removed.

**Comment by haampie** [comment]
@gartung any idea why `test_default_rpaths_create_install_default_layout` is failing? restarted the tests a couple times but it keeps failing.

**Comment by gartung** [comment]
Stage directory was not removed
https://github.com/spack/spack/blob/37731856395699956093d02a6ff64bf7cbf0f969/lib/spack/spack/test/conftest.py#L291

**Comment by haampie** [comment]
Sure, but it's failing spuriously, any idea why?

**Comment by gartung** [comment]
It's a test package so it creates it's own files on the fly. Maybe there is a timing issue with the cleanup of the stage directory.

File: lib/spack/spack/cmd/__init__.py
Comment (becker33) [inline lib/spack/spack/cmd/__init__.py]: This function should be `require_active_env`, because `require_env('add')` sounds like it's requiring an environment named "add", not requiring that an environment be active and passing a command name.

When we call this method, we should use the kwarg calling convention for clarity. `require_active_env(cmd_name='add')` is a lot clearer in this case.

File: lib/spack/spack/cmd/add.py
Comment (becker33) [inline lib/spack/spack/cmd/add.py]: See above, and same for all the similar instances.

File: lib/spack/spack/cmd/verify.py
Comment (becker33) [inline lib/spack/spack/cmd/verify.py]: You need to import `spack.environment` at the top here, and this goes for most of the files where you deleted it.

The only reason it's working without is because the name gets bound from importing `spack.store`, but you shouldn't rely on that because it's undefined python behavior.

**Review by becker33 — CHANGES_REQUESTED** [review]
I don't love the arrangement of methods in the cmd vs env modules, but I didn't love the old arrangement either so I'm not going to complain too much about that.

A couple minor issues.

File: lib/spack/spack/cmd/verify.py
Comment (becker33) [reply lib/spack/spack/cmd/verify.py]: @haampie I think this might be related to the test failure -- you can get some really inscrutable errors with improper imports.

Comment (haampie) [reply lib/spack/spack/cmd/verify.py]: Hm okay, let me try to fix the circular dependencies. Note that `def matching_spec_from_env` in `__init__.py` already used `spack.environment.<stuff>` without importing before this pr.

But I mean, this PR is intended to separate core domain stuff (spack.environment) from the interface layer (spack.cmd), so optimally there would not be any circular deps at all...

Comment (haampie) [reply lib/spack/spack/cmd/verify.py]: Couple of commits were necessary to fix circular imports, but then it's OK to add those `spack.environment` imports in spack.cmd.*. I've done it over the whole code base in 1ac0eb33d29ef06e45f6ccc102232a3b45adab0c, and made it consistently use `import spack.environment as ev` since this was already done in 55 out of 62 cases.

File: lib/spack/spack/util/__init__.py
Comment (becker33) [inline lib/spack/spack/util/__init__.py]: This probably belongs in llnl.util.lang (the same module as `dedupe` and `star`)

File: lib/spack/spack/cmd/__init__.py
Comment (becker33) [inline lib/spack/spack/cmd/__init__.py:491]: I don't fully understand why this method would be in spack.cmd... does that help with the circular import problem if you put it in `spack.environment` (and leave `find_environment` here because it uses argparse)?

**Comment by haampie** [comment]
@spackbot run pipeline

File: lib/spack/spack/environment.py
Comment (alalazo) [inline lib/spack/spack/environment.py]: Since this is a refactoring why not avoiding the `get_` prefix and call the function `active_environment`?

File: lib/spack/spack/installer.py
Comment (alalazo) [inline lib/spack/spack/installer.py]: I feel like this is the wrong place for this function if the aim was to have the spack core module not dependent on the cli. Here we assume knowledge on which arguments are passed in an `argparse` namespace.

File: lib/spack/spack/test/cmd/bootstrap.py
Comment (alalazo) [inline lib/spack/spack/test/cmd/bootstrap.py:10]: Why this change?

File: lib/spack/spack/architecture.py
Comment (alalazo) [inline lib/spack/spack/architecture.py]: Does this mean that this refactor introduces new circular dependencies at module creation? Is this due to some function that moved from one module to another?

File: lib/spack/spack/concretize.py
Comment (alalazo) [inline lib/spack/spack/concretize.py]: Why doing a local import rather than one at the top of the module?

File: lib/spack/spack/package.py
Comment (alalazo) [inline lib/spack/spack/package.py]: Why a local import?

**Review by alalazo — COMMENTED** [review]
I left a few comments. Overall I think this is an improvement over current state, and it catches a few places where we were missing imports. There's a function that I think is being moved to the wrong module and a few local imports that I would avoid if possible.

Style question to everyone. Do we want to consistently:
```python
import spack.environment as ev
```
everywhere? I have a slight preference for fully qualified names, but if we go with an alias everywhere I'm fine with it. It would be good if we can decide on that and have the decision checked automatically by style tests, if possible and easy to do.

File: lib/spack/spack/environment.py
Comment (haampie) [reply lib/spack/spack/environment.py]: Sounds good to me. Not sure what standard practices are in python-land, but we could as well drop the getter and rename _active_environment to active_environment and use it directly?

Comment (alalazo) [reply lib/spack/spack/environment.py]: I thought about that and it's certainly an option. If we use a function call we can later extend it with additional arguments with defaults without breaking the previous API.

File: lib/spack/spack/installer.py
Comment (haampie) [reply lib/spack/spack/installer.py]: Yeah, I tried to explain this in the commit message 89177317cc2b50e2845bf968507c8afa3a0ab0a5:

> The way I see it update_kwargs_from_args could have been a constructor
> for a PackageInstallArgs object, and they could have a named constructor
> like PackageInstallArgs::from_cli_args(args). So it makes sense to move
> this function to spack.installer, and with that the last dependency of
> spack.environment on spack.cmd is gone, as it should be.

Comment (haampie) [reply lib/spack/spack/installer.py]: I could move it out of installer.py, but I don't know where it should go instead :thinking: installer_args.py?

**Comment by haampie** [comment]
> Do we want to consistently:
> 
> ```python
> import spack.environment as ev
> ```
> 
> everywhere?

My reason for it was that we're doing it already in 88% of the cases:

```console
$ git checkout develop
$ grep '--include=*.py' 'import spack.environment as ev' -r lib/ | wc -l
55
$ grep '--include=*.py' 'import spack.environment' -r lib/ | wc -l
62
```

Edit: turns out python 2.7 struggles with alias'ed imports 6b8e412dcec00acd314ec48e942140121ffaf0b0, so `as ev` is used everywhere except in two cases now.

File: lib/spack/spack/installer.py
Comment (alalazo) [reply lib/spack/spack/installer.py]: Ok, I see that commit. My opinion is that the right solution is to have:

https://github.com/spack/spack/blob/b22728d55c126a0034f70c7331b6fabc4d6ad095/lib/spack/spack/environment.py#L1575

and similar functions not having the `args` argument. It should be on the caller of the interface to prepare the `install_args` appropriately.

File: lib/spack/spack/test/cmd/bootstrap.py
Comment (haampie) [reply lib/spack/spack/test/cmd/bootstrap.py:10]: See https://github.com/spack/spack/pull/25439#issuecomment-900906456

File: lib/spack/spack/installer.py
Comment (alalazo) [reply lib/spack/spack/installer.py]: Looking at:

https://github.com/spack/spack/blob/b22728d55c126a0034f70c7331b6fabc4d6ad095/lib/spack/spack/cmd/install.py#L197

it seems this is doable. Besides, `update_kwargs_from_args` is called _internally_ from `Environment.install_specs` which receives directly the `cli_args` and externally in the function above. So why not setting everything in `cmd/install.py` and just pass a dictionary after a single call to the function?

File: lib/spack/spack/architecture.py
Comment (haampie) [reply lib/spack/spack/architecture.py]: Hm, weird, by now it's no longer necessary... Maybe because I also deleted some imports elsewhere afterwards.

File: lib/spack/spack/concretize.py
Comment (haampie) [reply lib/spack/spack/concretize.py]: Because of Python 2:
```
$ SPACK_PYTHON=python2 spack --version
Traceback (most recent call last):
  File "/home/harmen/spack/bin/spack", line 96, in <module>
    import spack.main  # noqa
  File "/home/harmen/spack/lib/spack/spack/main.py", line 36, in <module>
    import spack.cmd
  File "/home/harmen/spack/lib/spack/spack/cmd/__init__.py", line 24, in <module>
    import spack.environment as ev
  File "/home/harmen/spack/lib/spack/spack/environment.py", line 20, in <module>
    import spack.concretize
  File "/home/harmen/spack/lib/spack/spack/concretize.py", line 36, in <module>
    import spack.environment as ev
AttributeError: 'module' object has no attribute 'environment'
```

There's a circularity between environment & concretize, but it's only an error on Python 2.7, not Python 3.x.

Comment (haampie) [reply lib/spack/spack/concretize.py]: Actually... when I drop `as ev` here, Python 2 does not complain :confused:.

File: lib/spack/spack/installer.py
Comment (haampie) [reply lib/spack/spack/installer.py]: I've removed `args` from `spack.environment.install_spec(...)` and `spack.environment.install_all(...)` and combined the `update_kwargs_from_args` in `spack.cmd.install`. That seems like a better solution indeed.

**Comment by haampie** [comment]
I think I've incorporated all the review comments now, and I've updated the PR description at the top with a summary of the changes.

File: lib/spack/spack/cmd/install.py
Comment (alalazo) [inline lib/spack/spack/cmd/install.py:329]: This looks much better to me now!

File: lib/spack/spack/environment.py
Comment (alalazo) [inline lib/spack/spack/environment.py:1484]: Definitely a better API imo without a dependency on an `argparse` namespace. Thanks for taking care of that!

File: lib/spack/spack/test/cmd/bootstrap.py
Comment (alalazo) [reply lib/spack/spack/test/cmd/bootstrap.py:10]: I think we can take care of understanding these two failures in a following PR.

**Review by alalazo — DISMISSED** [review]
LGTM but let's wait for @becker33 review. Should we wait on #22845 and #22720 to be merged before merging this PR, given  the number of files it touches?

File: lib/spack/spack/installer.py
Comment (alalazo) [reply lib/spack/spack/installer.py]: That's exactly what I had in mind. Thanks!

File: lib/spack/spack/concretize.py
Comment (alalazo) [reply lib/spack/spack/concretize.py]: Can you open an issue as soon as we merge this, to understand what's going on?

**Review by becker33 — CHANGES_REQUESTED** [review]
Mostly looks good, I agree the changes you made after Max's review are great improvements. only minor stuff left.

File: lib/spack/spack/cmd/__init__.py
Comment (haampie) [reply lib/spack/spack/cmd/__init__.py:491]: It's in `spack.cmd` because its only purpose is to create an error that shows you how to use the command if you didn't activate an env.

File: lib/spack/spack/util/__init__.py
Comment (haampie) [reply lib/spack/spack/util/__init__.py]: done

**Review by becker33 — APPROVED** [review]
I'm still not convinced on whether the methods are all in the perfect modules, but I don't think we should hold this up over it.

**Comment by haampie** [comment]
Thanks!

**Pull request merged**
